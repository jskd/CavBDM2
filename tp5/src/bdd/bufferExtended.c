/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
// Pour getline (standard GNU)
#define _GNU_SOURCE

#include "bufferExtended.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buffer {
  /// Valeur
  char* v;
  /// Taille
  size_t s;
  /// Nombre de valeur entrée
  size_t c;
  /// Taille des données
  size_t data_size;
};

/**
 * Creation d'un buffer contigue
 * @param[in] size      Taille du buffer
 * @param[in] data_size Taille des données
 */
struct buffer* buffer_create( size_t size , size_t data_size ) {
  struct buffer* buf= (struct buffer*) malloc(sizeof(struct buffer));
  buf->v= (char*) malloc( size * data_size );
  buf->s= size;
  buf->data_size= data_size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf) {
  free(buf->v);
  free(buf);
}


static char * _buffer_val_offset( const struct buffer* buf, int index) {
  return buf->v+index*buf->data_size;
}


/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[out] buf_dst  buffer de destination
 * @param[in] buf_src  buffer source
 * @param[in] src_index  index du buffer source à copier
 * @return    0   succès
 *            -1  erreur: buffer incompatible
 *            -2  erreur: buffer plein
 */
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index) {
  if(buf_dst->data_size != buf_src->data_size)
    return -1;

  return buffer_put(buf_dst, _buffer_val_offset(buf_src, src_index));
}

char buffer_put(struct buffer* buf, const void * data) {
  if(buf->c >= buf->s)
    return -1;

  memcpy ( _buffer_val_offset(buf, buf->c), data, buf->data_size );
  buf->c++;

  return 0;
}



/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buffer* buf= buffer_create( buffer_size, data_lenght );

  buffer_read_file_from_descriptor(file, buf);

  fclose(file);

  return buf;
}



void buffer_flush(struct buffer* buf) {
  buf->c= 0;
}


void buffer_read_file_from_descriptor(FILE* file, struct buffer* buf)
{
  buffer_flush(buf);
  rewind(file);

  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, file)) != -1) {
    buffer_put(buf, line);
  }
}




/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, const struct buffer* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  buffer_write_file_from_descriptor(file, buf);

  fclose(file);
  return 0;
}



void buffer_write_file_from_descriptor(FILE* file, const struct buffer* buf) {
  for(int i=0; i<buf->c; i++)
    fprintf(file, "%.*s\n", (int) buf->data_size, _buffer_val_offset(buf, i));
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buffer_quicksort(struct buffer* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c * buf->data_size);
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf)
{
  return buf->c;
}

size_t buffer_size(const struct buffer* buf)
{
  return buf->s;
}
int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b) {
  return memcmp( _buffer_val_offset(buf_a, index_a),
     _buffer_val_offset(buf_b, index_b), buf_a->data_size);
}
