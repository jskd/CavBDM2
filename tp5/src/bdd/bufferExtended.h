/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#ifndef SRC_BDD_BUFFER_EXTENDED_H
#define SRC_BDD_BUFFER_EXTENDED_H

#include <stdio.h>

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct buffer;

/**
 * Creation d'un buffer contigue
 * @param[in] size      Taille du buffer
 * @param[in] data_size Taille des données
 */
struct buffer* buffer_create( size_t size , size_t data_size );

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf);


char buffer_put(struct buffer* buf, const void * data);

/**
 * @brief Quick sort buffer
 *
 * @param buf Trie le buffer en entrée
 */
void buffer_quicksort(struct buffer* buf);

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 * @param[in]  data_lenght Taille des données ajouté
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght);


/**
 * @brief Remplis le buffer du contenu du fichier
 *        N'alloue pas de buffer et vide de buffer avant d'effectuer l'operation
 *
 * @param[in]  File_name Fichier
 * @param[out] buffer
 */
void buffer_read_file_from_descriptor(FILE* fp, struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, const struct buffer*);



void buffer_write_file_from_descriptor(FILE* file, const struct buffer* buf);

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 */
char buffer_val(const struct buffer* buf, int index);

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf);

size_t buffer_size(const struct buffer* buf);
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index);
void buffer_flush(struct buffer* buf);

int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b);
#endif
