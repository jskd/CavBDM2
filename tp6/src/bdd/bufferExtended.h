/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#ifndef SRC_BDD_BUFFER_EXTENDED_H
#define SRC_BDD_BUFFER_EXTENDED_H

#include <stdio.h>

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct buffer;

typedef enum {
  BUFFER_CHARACTERS,
  BUFFER_DECIMALS
} buffer_file_mode;


/**
 * Creation d'un buffer contigue
 * @param[in] size Taille du buffer
 * @param[in] num  Taille des données en octet
 * @param[in] mode Mode de fonctionnement des sortie fichier
 */
struct buffer* buffer_create( size_t size , size_t num, buffer_file_mode mode );

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf);


char buffer_put(struct buffer* buf, const void * data);

/**
 * @brief Quick sort buffer
 *
 * @param buf Trie le buffer en entrée
 */
void buffer_quicksort(struct buffer* buf);

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 * @param[in]  data_lenght Taille des données ajouté
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght, buffer_file_mode mode);


/**
 * @brief Remplis le buffer du contenu du fichier
 *        N'alloue pas de buffer et vide de buffer avant d'effectuer l'operation
 *
 * @param[in]  File_name Fichier
 * @param[out] buffer
 */
void buffer_read_file_from_descriptor(FILE* fp, struct buffer* buf);

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, const struct buffer*);


/**
 * Ecrit le contenu d'un buffer dans un fichier
 */
void buffer_write_file_from_descriptor(FILE* file, const struct buffer* buf);

/**
 * @brief Copie la valeur d'une entrée d'un buffer dans le block memoire
 *        destination
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 * @param[out] destination
 */
void buffer_get(const struct buffer* buf, int index,  void* destination);

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf);

/**
 * Retourne la taille du buffer
 */
size_t buffer_size(const struct buffer* buf);

/**
 * Copy la src_index valeur de buf_src et la stoque dans buf_dst (à la dernière position)
 *
 * @param[out] buf_dst  buffer de destination
 * @param[in] buf_src  buffer source
 * @param[in] src_index  index du buffer source à copier
 * @return    0   succès
 *            -1  erreur: buffer incompatible
 *            -2  erreur: buffer plein
 */
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index);

/**
 * Vide le buffer
 */
void buffer_flush(struct buffer* buf);

/**
 *  Compares 2 buffer value:
 *  <0 lower value in buf_a than in buf_b
 *  0 equal
 *  >0 greater value in buf_a than in buf_b
 */
int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b);

/**
 * Retourne si le buffer est complet
 * @return    1   complet
 *            0 non complet
 */
char buffer_isFull(const struct buffer* buf);

/**
 * Retourne si le buffer est vide
 * @return    0 vide
 *            1 contien des données
 */
char buffer_isEmpty(const struct buffer* buf);


/**
 * Affiche les valeurs
 */
void buffer_printValue(const struct buffer* buf);

/**
 * Retourne le mode de fonctionnement du buffer
 */
int buffer_mode(const struct buffer* buf);

/**
 * Retourne la taille en octet de chaque données
 */
size_t buffer_datasize(const struct buffer* buf);

#endif
