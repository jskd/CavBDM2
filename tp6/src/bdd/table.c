/**
 * TP n°: 6
 *
 * Titre du TP :
 *
 * Date : 21/10/17
 *
 * Nom : Lefranc
 * Prenom : Joaquim
 * email : lefrancjoaquim@gmail.com
 *
 * Nom : Skoda
 * Prenom : Jérôme
 * email : contact@jeromeskoda.fr
 *
 * Remarques :
 */
#include "table.h"
#include "buffer.h"
#include "bucket.h"
#include <limits.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define BUCKET_FILE_EXTENSION ".txt"
/**
 * @brief Structure de donnée repésentant un buffer
 */
struct table {
  /// bucket
  struct bucket** b;
  /// n bucket
  size_t n_bucket;
};

static int _table_hash(const struct table* tab, short key) {
  return key % tab->n_bucket;
}

struct table* table_create(size_t n_bucket, const char* directory) {
  struct table* tab= (struct table*) malloc(sizeof(struct table));
  tab->b= (struct bucket**) malloc(n_bucket * sizeof(struct bucket*) );
  mkdir(directory, 0777);
  for(int indexBucket=0; indexBucket<n_bucket; indexBucket++) {
    char filename[PATH_MAX];
    sprintf(filename, "%s/%d.bucket.txt", directory, indexBucket);
    tab->b[indexBucket]= bucket_create(filename);
  }
  tab->n_bucket= n_bucket;
  return tab;
}

void table_putBuffer(struct table* tab, const struct buffer* buf) {
  if(buffer_mode(buf) == BUFFER_DECIMALS
    && buffer_datasize(buf) == sizeof(short))
  {
    for(int index_buf=0; index_buf< buffer_count(buf); index_buf++) {
      short key=0;
      char str[100];
      sprintf(str, "%d", key);
      bucket_puts( tab->b[ _table_hash(tab, key) ], str);
    }
  }
  else
  {
    printf("Erreur, table put non implementé pour ce type de buffer\n");
  }
}

void table_storeBucketInBuffer(struct table* tab, int index, struct buffer* buf)
{
  if(index > 0 && index < tab->n_bucket)
    buffer_read_file_from_descriptor(  bucket_getFile(tab->b[index]), buf);
}
