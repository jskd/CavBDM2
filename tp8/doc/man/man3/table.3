.TH "table" 3 "Vendredi 1 Décembre 2017" "CavBDM2 - BDD" \" -*- nroff -*-
.ad l
.nh
.SH NAME
table \- Structure de donnée repésentant un buffer\&.  

.SH SYNOPSIS
.br
.PP
.SS "Attributs publics"

.in +1c
.ti -1c
.RI "struct \fBbucket\fP ** \fBb\fP"
.br
.RI "bucket "
.ti -1c
.RI "size_t \fBn_bucket\fP"
.br
.RI "n bucket "
.ti -1c
.RI "int \fBwrite_counter\fP"
.br
.RI "bench "
.in -1c
.SH "Description détaillée"
.PP 
Structure de donnée repésentant un buffer\&. 
.SH "Documentation des données membres"
.PP 
.SS "struct \fBbucket\fP** table::b"

.PP
bucket 
.SS "size_t table::n_bucket"

.PP
n bucket 
.SS "int table::write_counter"

.PP
bench 

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour CavBDM2 - BDD à partir du code source\&.
