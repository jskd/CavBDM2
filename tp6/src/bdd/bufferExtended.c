/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
// Pour getline (standard GNU)
#define _GNU_SOURCE

#include "bufferExtended.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buffer {
  /// Valeur
  char* v;
  /// Taille
  size_t s;
  /// Nombre de valeur entrée
  size_t c;
  /// Taille des données
  size_t data_size;
  // mode des sortie fichier
  buffer_file_mode mode;
  /// compteur de Lecture (en octet)
  int read_counter;
  /// compteur d'écriture (en octet)
  int write_counter;
};

// See Header
struct buffer* buffer_create( size_t size , size_t num, buffer_file_mode mode ){
  struct buffer* buf= (struct buffer*) malloc(sizeof(struct buffer));
  buf->v= (char*) malloc( size * num );
  buf->s= size;
  buf->data_size= num;
  buf->c=0;
  buf->mode = mode;
  buf->read_counter = 0;
  buf->write_counter = 0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buffer_destroy(struct buffer* buf) {
  free(buf->v);
  free(buf);
}


static char * _buffer_val_offset( const struct buffer* buf, int index) {
  return buf->v+index*buf->data_size;
}

static void _printValue(FILE* stream, struct buffer* buf, int index) {
  if(buf->mode == BUFFER_CHARACTERS) {
    fprintf(stream, "%.*s\n", (int) buf->data_size, _buffer_val_offset(buf, index));
  } else if(buf->mode == BUFFER_DECIMALS) {
    long long value=0;
    buffer_get(buf, index, &value);
    switch(buf->data_size) {
      case 1:
        fprintf(stream, "%d\n",  (char) value);
        break;
      case 2:
        fprintf(stream, "%d\n",  (short) value);
        break;
      case 4:
        fprintf(stream, "%d\n",  (int) value);
        break;
      default:
        printf("Print value decimimal lenght %ld not implemented\n", buf->data_size);
        return;
        break;
    }
  }
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[out] buf_dst  buffer de destination
 * @param[in] buf_src  buffer source
 * @param[in] src_index  index du buffer source à copier
 * @return    0   succès
 *            -1  erreur: buffer incompatible
 *            -2  erreur: buffer plein
 */
char buffer_put_cpy(struct buffer* buf_dst, const struct buffer* buf_src, int src_index) {
  if(buf_dst->data_size != buf_src->data_size)
    return -1;

  return buffer_put(buf_dst, _buffer_val_offset(buf_src, src_index));
}

char buffer_put(struct buffer* buf, const void * data) {
  if(buf->c >= buf->s)
    return -1;

  memcpy ( _buffer_val_offset(buf, buf->c), data, buf->data_size );
  buf->c++;

  return 0;
}

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buffer* buffer_read_file(const char* file_name, size_t buffer_size, size_t data_lenght, buffer_file_mode mode) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buffer* buf= buffer_create( buffer_size, data_lenght, mode );

  buffer_read_file_from_descriptor(file, buf);

  fclose(file);

  return buf;
}



void buffer_flush(struct buffer* buf) {
  buf->c= 0;
}


void buffer_read_file_from_descriptor(FILE* file, struct buffer* buf)
{
  buffer_flush(buf);
  rewind(file);

  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, file)) != -1) {
    if(buf->mode == BUFFER_CHARACTERS) {
      buffer_put(buf, line);
    }
    else if(buf->mode == BUFFER_DECIMALS) {
      int value= atoi(line);
      buffer_put(buf, &value);
    }
    buf->read_counter++;
  }
}

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char buffer_write_file(const char* file_name, struct buffer* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  buffer_write_file_from_descriptor(file, buf);

  fclose(file);
  return 0;
}



void buffer_write_file_from_descriptor(FILE* file, struct buffer* buf) {
  for(int index=0; index<buf->c; index++) {
    _printValue(file, buf, index);
    if(file != stderr && file != stdout)
      buf->write_counter++;
  }
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buffer_quicksort(struct buffer* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buffer_dump(const struct buffer* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c * buf->data_size);
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buffer_count(const struct buffer* buf)
{
  return buf->c;
}

size_t buffer_size(const struct buffer* buf)
{
  return buf->s;
}
int buffer_cmp(const struct buffer* buf_a, int index_a, const struct buffer* buf_b, int index_b) {
  return memcmp( _buffer_val_offset(buf_a, index_a),
     _buffer_val_offset(buf_b, index_b), buf_a->data_size);
}

char buffer_isFull(const struct buffer* buf) {
  return buf->s == buf->c;
}

char buffer_isEmpty(const struct buffer* buf) {
  return buf->c == 0;
}

void buffer_get(const struct buffer* buf, int index, void* destination) {
  memcpy ( destination,  _buffer_val_offset(buf, index), buf->data_size );
}

/**
 * Affiche les valeurs
 */
void buffer_printValue(struct buffer* buf) {
  for(int index=0; index< buf->c; index++) {
    _printValue(stdout, buf, index);
  }
}

/**
 * Retourne le mode de fonctionnement du buffer
 */
int buffer_mode(const struct buffer* buf) {
  return buf->mode;
}

/**
 * Retourne la taille en octet de chaque données
 */
size_t buffer_datasize(const struct buffer* buf) {
  return buf->data_size;
}


int buffer_get_read_stat(const struct buffer* buf) {
  return  buf->read_counter;
}

int buffer_get_write_stat(const struct buffer* buf) {
  return  buf->write_counter;
}


void buffer_fprint_stat(FILE* stream, const struct buffer* buf) {
  fprintf(stream, "- Read in file: %d lines\n", buf->read_counter);
  fprintf(stream, "- Write in file: %d lines\n", buf->write_counter);
}
