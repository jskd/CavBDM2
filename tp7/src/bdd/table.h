/**
 * TP n°: 6
 *
 * Titre du TP :
 *
 * Date : 21/10/17
 *
 * Nom : Lefranc
 * Prenom : Joaquim
 * email : lefrancjoaquim@gmail.com
 *
 * Nom : Skoda
 * Prenom : Jérôme
 * email : contact@jeromeskoda.fr
 *
 * Remarques :
 */
#ifndef SRC_BDD_TABLE_H
#define SRC_BDD_TABLE_H

#include <stdio.h>
#include "bufferExtended.h"

struct table;

struct table* table_create(size_t bucket, const char* directory);

void table_insertbuffer(struct table* tab, const struct buffer* buf);

void table_putBuffer(struct table* tab, const struct buffer* buf);

void table_destroy(struct table* tab) ;
size_t table_get_n_bucket(const struct table* tab);

int table_get_write_stat(const struct table* tab);
void table_fprint_stat(FILE* stream, const struct table* tab);

struct diskReader* create_disk_from_bucket(const struct table* tab, int indexBucket) ;
#endif
