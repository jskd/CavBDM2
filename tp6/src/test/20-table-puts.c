#include "../bdd/bufferExtended.h"
#include "../bdd/table.h"

int main(int argc, char** argv){

  struct buffer* buf= buffer_read_file("res/test/20-buffer-pair-impaire.txt",
    10, sizeof(short), BUFFER_DECIMALS);

  struct table* tab= table_create(2, "res/test/20-bucket");

  table_putBuffer(tab, buf);

  table_storeBucketInBuffer(tab, 0, buf);

  printf("bucket 0\n");
  buffer_printValue(buf);

  table_storeBucketInBuffer(tab, 1, buf);

  printf("bucket 1\n");
  buffer_printValue(buf);

  buffer_destroy(buf);

}
