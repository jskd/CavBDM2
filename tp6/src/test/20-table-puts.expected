bucket 0
22
4444
-66
-8888
bucket 1
1
333
-5
-777
